### DISCUSSION TITLE
Is there any way to mark some but not all textures as external?

### DISCUSSION THREAD

**Discussion by looeee** [donmccurdy/glTF-Transform#712] [Q&A] [answered]
I have a bunch of similar models that I am processing with gltf-transform. Nearly all of them use shared textures that display a logo and a QR code alongside unique textures for normal maps, roughness, color, AO, and so on. 

As far as I understand it, if the logo and QR code are embedded, they won't get picked up by three.js cache - for this to work they need to be external. So if they are embedded I will be recreating identical textures once for each model - potentially twenty or thirty times - and I'd like to avoid that. 

For example, one model might have these two textures where the QR code is shared and the AO is unique:

```
  "images": [
    { "uri": "tex_default_QR_code_a.png" },
    { "uri": "tex_xxx_front_AO.jpg" },
  ],
```

I would like to adjust this like so:

```
  "images": [
    { "uri": "shared_textures/tex_default_QR_code_a.png" },
    { "uri": "tex_xxx_front_AO.jpg" },
  ],
```

Following this, I would like to generate the .glb which references any textures in `shared_textures` as external, and embeds all other textures. Is there any way to do this currently?

**Comment by looeee** [comment]
Two alternatives I am considering: 

1. Put all textures in a external `assets/textures` folder and set `GLTFLoader.path = /assets/textures`. There will probably be 100 or so textures, I'm not sure what the performance cost of make them external would be. Besides that, this communal textures folder will be messy and we'll have to get stricter with texture naming conventions, although that's not a deal breaker. 
2. I am working with an artist to iterate on the models, and it's much easier for her to give them to me separately. However, I could pre-process to bundle them all into a single model and presumably that would dedup the shared textures. 

With 1. I am mainly worried about performance. 

The problem with 2. is that we will always need to load all models. Currently that's ok but in the future we may want to improve load times by  only loading models as they are required and that won't be possible if they are all in one big file.

**Reply by donmccurdy** [reply]
> Following this, I would like to generate the .glb which references any textures in shared_textures as external, and embeds all other textures. Is there any way to do this currently?

Not currently — textures are written either externally (.gltf) or embedded (.glb) and there are no per-texture settings to change that today.

> 1. ... There will probably be 100 or so textures, I'm not sure what the performance cost of make them external would be.

It's not obvious to me that this will have any performance cost — it may even improve download times to have a larger number of smaller files. Or at least I've heard that argument made, I have not seen numbers either way.

> 2. ... However, I could pre-process to bundle them all into a single model and presumably that would dedup the shared textures.

This would dedup the textures, yes. But if it doesn't match with how/when you actually want to load the models, I suspect that's not worth it.

> As far as I understand it, if the logo and QR code are embedded, they won't get picked up by three.js cache...

Yeah, the network-level THREE.Cache won't be able to help with these. The GLTFLoader plugin API would probably let you replace similarly named textures with the same THREE.Texture instance, but any download size is still there, and then that's custom runtime code to maintain. Not sure I'd want to take that approach myself.

***

My hunch would be that (1) and a consistent naming convention (or folder prefix) would probably be the way to go here. If a method of embedding specific textures and exporting others gets added in the future, you'll need those identifiers on the textures anyway.

**Reply by looeee** [reply]
> textures are written either externally (.gltf) or embedded (.glb) and there are no per-texture settings to change that today.

OK, thanks for the confirmation. I figured this was the case but I was hoping I had missed something. Is this a tooling limitation or does it come from the glTF spec itself? 

> It's not obvious to me that this will have any performance cost

I've also heard that but likewise not seen any numbers. However I'm willing to accept it on faith for now. Differences are probably insignificant in any case. 

> My hunch would be that (1) and a consistent naming convention (or folder prefix) would probably be the way to go here

Yeah the textures are already well named so that shouldn't require any additional work. We'll go with this, thanks for you help.

**Reply by looeee** [reply]
I came across this thread just now - in case anyone else comes across this discussion it's relevant. 

https://github.com/KhronosGroup/glTF/issues/1117

Especially [@javagl's  comment](https://github.com/KhronosGroup/glTF/issues/1117#issuecomment-336633574). It seems like no-one has followed up on that since. 

> A broader question would be about the form or feature set of this tool. I think that there may be different levels of granularity to answer the question of what should be exported how:
> 1. Just convert a whole glTF+BIN+Images into a GLB
> 2. Convert a glTF+BIN into a GLB, but keep the images as external references
> 3. Convert a glTF+BIN+Images into a GLB, but keep some specific (user-selected) images as external references

I guess the tool being discussed here is what eventually became gltf-transform (and maybe gltf-pipeline too). 

However as far as I can see, gltf-transform currently only supports these two options? 

1. everything embedded (.glb)
2. nothing embedded (.gltf + bin + images)

**Reply by donmccurdy** [reply, marked as answer]
> Is this a tooling limitation or does it come from the glTF spec itself?

No limitation in the spec. I think it's just an uncommon enough goal that tools (including mine) have not added options to export embedded vs. external textures with a "mix-and-match" approach.

> However as far as I can see, gltf-transform currently only supports these two options?

That's correct — there's some flexibility in assigning binary data to multiple .bin files. But I haven't implemented any way to write .glb files that reference external data.

**Reply by looeee** [reply]
OK cool, that's fine. However I would love to see this option added at some point:

> Convert a glTF+BIN into a GLB, but keep the images as external references

Consider it a very low priority feature request :)

**Reply by donmccurdy** [reply]
Feel free to open a ticket! I'm open to the idea but not sure what the API should be. A property like `texture.setExternal(true)` would conflict with other ways of specifying glTF vs GLB, so I think can't be a property of the texture itself, but rather something configured on the I/O instance...
